doc/man/man1/sacct.1
.TH sacct "1" "Slurm Commands" "August 2022" "Slurm Commands"

.SH "NAME"
sacct \- displays accounting data for all jobs and job steps in the
Slurm job accounting log or Slurm database

.SH "SYNOPSIS"
\fBsacct\fR [\fIOPTIONS\fR...]

.SH "DESCRIPTION"
.PP
Accounting information for jobs invoked with Slurm are either logged
in the job accounting log file or saved to the Slurm database, as
configured with the AccountingStorageType parameter.
.PP
The \f3sacct\fP command displays job accounting data stored in the job
accounting log file or Slurm database in a variety of forms for your
analysis.  The \f3sacct\fP command displays information on jobs, job
steps, status, and exitcodes by default.  You can tailor the output
with the use of the \f3\-\-format=\fP option to specify the fields to
be shown.
.PP
Job records consist of a primary entry for the job as a whole as well as
entries for job steps. The Job Launch page has a more detailed description
of each type of job step.
<https://slurm.schedmd.com/job_launch.html#job_record>
.PP
For the root user, the \f3sacct\fP command displays job accounting
data for all users, although there are options to filter the output to
report only the jobs from a specified user or group.
.PP
For the non\-root user, the \f3sacct\fP command limits the display of
job accounting data to jobs that were launched with their own user
identifier (UID) by default.  Data for other users can be displayed
with the \f3\-\-allusers\fP, \f3\-\-user\fP, or \f3\-\-uid\fP options.
.PP
Elapsed time fields are presented as
[days\-]hours:minutes:seconds[.microseconds]. Only 'CPU' fields will
ever have microseconds.
.PP
The default input file is the file named in the
\f3AccountingStorageLoc\fP parameter in slurm.conf.

\fBNOTE\fR: If designated, the slurmdbd.conf option PrivateData may further
restrict the accounting data visible to users which are not
SlurmUser, root, or a user with AdminLevel=Admin. See the
slurmdbd.conf man page for additional details on restricting
access to accounting data.

\fBNOTE\fR: The contents of Slurm's database are maintained in lower case.
This may result in some \f3sacct\fP output differing from that of other Slurm
commands.

\fBNOTE\fR: Much of the data reported by \f3sacct\fP has been generated by
the \f2wait3()\fP and \f2getrusage()\fP system calls. Some systems
gather and report incomplete information for these calls;
\f3sacct\fP reports values of 0 for this missing data. See your systems
\f2getrusage (3)\fP man page for information about which data are
actually available on your system.

.SH "OPTIONS"

.TP "10"
\fB\-A\fR, \fB\-\-accounts\fR=<\fIaccount_list\fR>
Displays jobs when a comma separated list of accounts are given as the
argument.
.IP

.TP
\fB\-L\fR, \fB\-\-allclusters\fR
Display jobs ran on all clusters. By default, only jobs ran on the
cluster from where \fBsacct\fR is called are displayed.
.IP

.TP
\fB\-X\fR, \fB\-\-allocations\fR
Only show statistics relevant to the job allocation itself, not taking steps
into consideration.

\fBNOTE\fR: Without including steps, utilization statistics for job
allocation(s) will be reported as zero.
.IP

.TP
\fB\-a\fR, \fB\-\-allusers\fR
Displays all users' jobs when run by user root or if \fBPrivateData\fP is not
configured to \fBjobs\fP.
Otherwise display the current user's jobs
.IP

.TP
\fB\-x\fR, \fB\-\-associations\fR=<\fIassoc_list\fR>
Displays the statistics only for the jobs running under the
association ids specified by the \fBassoc_list\fR operand, which is a
comma\-separated list of association ids.  Space characters are not
allowed in the \fBassoc_list\fR. Default is all associations\&.
.IP

.TP
\fB\-B\fR, \fB\-\-batch\-script\fR
This option will print the batch script of job if the job used one. If the job
didn't have a script 'NONE' is output.
.br
\fBNOTE\fR: AccountingStoreFlags=job_script is required for this.
.br
\fBNOTE\fR: Requesting specific job(s) with '\-j' is required for this.
.IP

.TP
\fB\-b\fR, \fB\-\-brief\fR
Displays a brief listing consisting of JobID, State, and ExitCode.
.IP

.TP
\fB\-M\fR, \fB\-\-clusters\fR=<\fIcluster_list\fR>
Displays the statistics only for the jobs started on the clusters
specified by the \fIcluster_list\fR operand, which is a
comma\-separated list of clusters.  Space characters are not allowed
in the \fIcluster_list\fR.
A value of '\fIall\fR' will query to run on all clusters.
The default is current cluster you are executing the \fBsacct\fR command on or
all clusters in the federation when executed on a federated cluster.
This option implicitly sets the \fB\-\-local\fR option.
.IP

.TP
\fB\-c\fR, \fB\-\-completion\fR
Use job completion data instead of job accounting.  The \fBJobCompType\fR
parameter in the slurm.conf file must be defined to a non\-none option.
Does not support federated cluster information (local data only).
.IP

.TP
\fB\-C\fR, \fB\-\-constraints\fR=<\fIconstraint_list\fR>
Comma separated list to filter jobs based on what constraints/features the job
requested.  Multiple options will be treated as 'and' not 'or', so the job would
need all constraints specified to be returned not one or the other.
.IP

.TP
\fB\-\-delimiter\fR=<\fIcharacters\fR>
ASCII characters used to separate the fields when specifying
the \fB\-p\fR or \fB\-P\fR options. The default delimiter
is a '|'. This option is ignored if \fB\-p\fR or \fB\-P\fR options
are not specified.
.IP

.TP
\fB\-D\fR, \fB\-\-duplicates\fR
If Slurm job ids are reset, some job numbers will probably appear more
than once in the accounting log file but refer to different jobs.
Such jobs can be distinguished by the "submit" time stamp in the data
records.
.IP

When data for specific jobs are requested with the \-\-jobs option,
\fBsacct\fR returns the most recent job with that number. This
behavior can be overridden by specifying \-\-duplicates, in which case
all records that match the selection criteria will be returned.
.IP

\fBNOTE\fR: Revoked federated sibling jobs are hidden unless the
\fB\-\-duplicates\fR option is specified.
.IP

.TP
\fB\-E\fR, \fB\-\-endtime\fR=<\fIend_time\fR>
Select jobs in any state before the specified time.  If states are
given with the \-s option return jobs in this state before this period.
See the \fBDEFAULT TIME WINDOW\fR for more details.

Valid time formats are:
.br
HH:MM[:SS][AM|PM]
.br
MMDD[YY][\-HH:MM[:SS]]
.br
MM.DD[.YY][\-HH:MM[:SS]]
.br
MM/DD[/YY][\-HH:MM[:SS]]
.br
YYYY\-MM\-DD[THH:MM[:SS]]
.br
today, midnight, noon, fika (3 PM), teatime (4 PM)
.br
now[{+|\-}\fIcount\fR[seconds(default)|minutes|hours|days|weeks]]
.IP

.TP
\fB\-\-env\-vars\fR
This option will print the running environment of a batch job, otherwise 'NONE'
is output.
.br
\fBNOTE\fR: AccountingStoreFlags=job_env is required for this.
.br
\fBNOTE\fR: Requesting specific job(s) with '\-j' is required for this.
.IP

.TP
\fB\-\-federation\fR
Show jobs from the federation if a member of one.
.IP

.TP
\fB\-f\fR, \fB\-\-file\fR=<\fIfile\fR>
Causes the \f3sacct\fP command to read job accounting data from the
named \f2file\fP instead of the current Slurm job accounting log
file. Only applicable when running the jobcomp/filetxt plugin. Setting this flag
implicitly enables the \-c flag.
.IP

.TP
\fB\-F\fR, \fB\-\-flags\fR=<\fIflag_list\fR>
Comma separated list to filter jobs based on what various ways the jobs were
handled.  Current flags are SchedSubmit, SchedMain, SchedBackfill.  These
particular options describe the scheduler that started the job.
.IP

.TP
\fB\-o\fR, \fB\-\-format\fR
Comma separated list of fields. (use "\-\-helpformat" for a list of
available fields).

\fBNOTE\fR: When using the format option for listing various fields you can put
a %NUMBER afterwards to specify how many characters should be printed.

e.g. format=name%30 will print 30 characters of field name right
justified.  A %\-30 will print 30 characters left justified.

When set, the SACCT_FORMAT environment variable will override the
default format.  For example:

SACCT_FORMAT="jobid,user,account,cluster"
.IP

.TP
\fB\-g\fR, \fB\-\-gid\fR=, \fB\-\-group\fR=<\fIgid_or_group_list\fR>
Displays the statistics only for the jobs started with the GID
or the GROUP specified by the \fIgid_list\fR or the \fIgroup_list\fR operand,
which is a comma\-separated list.  Space characters are not allowed.
Default is no restrictions.
.IP

.TP
\fB\-h\fR, \fB\-\-help\fR
Displays a general help message.
.IP

.TP
\fB\-e\fR, \fB\-\-helpformat\fR
Print a list of fields that can be specified with the \fB\-\-format\fR option.
.IP

.RS
.PP
.nf
.ft 3
Fields available:

Account             AdminComment        AllocCPUS           AllocNodes
AllocTRES           AssocID             AveCPU              AveCPUFreq
AveDiskRead         AveDiskWrite        AvePages            AveRSS
AveVMSize           BlockID             Cluster             Comment
Constraints         ConsumedEnergy      ConsumedEnergyRaw   Container

RELEASE_NOTES
metastack2.1版本发布说明

重要提示：
metastack 是基于 Slurm 的功能增强和性能优化的版本，metastack2.1基于Slurm-22.05.8版本进行开发。

如果使用 slurmdbd（Slurm 数据库守护进程），则必须先更新此版本。

注意：如果使用主备DBD，则必须首先启动主数据库才能进行任何数据库转换。

metastack2.1的slurmdbd将与metastack2.0及以上版本、Slurm-20.11及以上版本的Slurm守护程序兼容。不需同时更新所有集群，但是在更新其他使用slurmdbd的集群之前，需首先更新slurmdbd并使其运行。

metastack2.1可以从metastack2.0及以上版本、Slurm-20.11或Slurm-21.08版本直接升级到该版本，不会丢失作业或其他状态信息。直接从早期版本的 Slurm 升级将导致状态信息的丢失。

从其他版本升级到metastack2.1时，所有SPANK插件必须重新编译。


要点
=====================================================================
-- 支持作业运行信息采集，如CPU利用率、内存使用情况等
-- 支持异常负载作业识别与监控
-- 支持计算节点随机调度，均衡计算节点使用率
-- 支持分区节点资源动态调整
-- 支持调度/查询服务分离，应对服务高负载场景
-- 支持低版本兼容，兼容metastack低版本与slurm官方版本
-- 优化账号/用户相关的添加、查询性能问题
-- 优化单位时间内用户请求流量控制功能
-- 优化部分RPC高负载场景
-- 优化分区异构资源调度支持
-- 优化作业提交性能
-- 优化大批量小作业场景下作业吞吐效率
-- 优化srun创建作业步超时场景处理
-- 修复bug若干


配置文件更改（详见man page帮助手册）
=====================================================================
-- slurm.conf
    - JobAcctGatherFrequency新增配置子项time_window、avecpuutil、collect_step、disable_job_monitor
    - SlurmctldParameters新增配置子项rl_enable、rl_retry_max_period、node_borrow_interval、step_pendlist_dedup、step_timeout_enhance
    - DebugFlags新增配置子项Assoc
    - 新增配置项CacheQuery、CacheQueryPort、CacheDupInterval
    - 新增配置项DownNodeToReg、HealthCheckCarryNode
    - 分区配置新增配置子项RBN
    - 分区配置新增配置子项HetPart
    - 分区配置新增配置子项StandbyNodes、StandbyNodeParameters，其中StandbyNodeParameters包含配置子项nodes_can_borrow、offline_node_state
-- acct_gather.conf
    - 新增配置项ProfileInfluxDBDefault，可配置task,jobmonitor
    - 新增配置项ProfileInfluxDBWorkdir


命令更改（详见man page帮助手册）
=====================================================================
-- sinfo
    - 新增--borrow参数
    - 新增--cache参数
-- squeue
    - 新增--cache参数
-- scontrol 
    - 分区查询新增--borrow参数
    - 作业/分区/节点/作业步查询新增--cache参数
-- sacctmgr 
    - 查询新增borrow参数
    - 新增-y、-z参数，分别对应环境变量SACCTMGR_NO_PARENTLIMITS、SACCTMGR_NO_SORT
-- sbatch
    - 新增--job-monitor参数，参数可配置子项time_window、avecpuutil、collect_step
-- sstat
    - 新增-d参数
-- sjinfo
    - 新增命令sjinfo，可选参数-j、-e、-o、-m等


src/sacct/sacct.c
/*****************************************************************************\
 *  sacct.c - job accounting reports for Slurm's jobacct/log plugin
 *****************************************************************************
 *
 *  Copyright (C) 2008-2009 Lawrence Livermore National Security.
 *  Copyright (C) 2005 Hewlett-Packard Development Company, L.P.
 *
 *  This file is part of Slurm, a resource management program.
 *  For details, see <https://slurm.schedmd.com/>.
 *  Please also read the included file: DISCLAIMER.
 *
 *  Slurm is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  In addition, as a special exception, the copyright holders give permission
 *  to link the code of portions of this program with the OpenSSL library under
 *  certain conditions as described in each individual source file, and
 *  distribute linked combinations including the two. You must obey the GNU
 *  General Public License in all respects for all of the code used other than
 *  OpenSSL. If you modify file(s) with this exception, you may extend this
 *  exception to your version of the file(s), but you are not obligated to do
 *  so. If you do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source files in
 *  the program, then also delete it here.
 *
 *  Slurm is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Slurm; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
\*****************************************************************************/

#include "src/common/slurm_auth.h"

#include "sacct.h"

/*
 * Globals
 */
sacct_parameters_t params;
#ifdef  __METASTACK_OPT_PRINT_COMMAND
print_field_t fields[] = {
	{10, "Account", print_fields_str, PRINT_ACCOUNT ,0},
	{15, "AdminComment", print_fields_str, PRINT_ADMIN_COMMENT,0},
	{10, "AllocCPUS", print_fields_uint, PRINT_ALLOC_CPUS, 0},
	{10, "AllocNodes", print_fields_str, PRINT_ALLOC_NODES, 0},
	{10, "AllocTRES", print_fields_str, PRINT_TRESA, 0},
	{7,  "AssocID", print_fields_uint, PRINT_ASSOCID, 0},
	{10, "AveCPU", print_fields_str, PRINT_AVECPU, 0},
	{10, "AveCPUFreq", print_fields_str, PRINT_ACT_CPUFREQ, 0},
	{14, "AveDiskRead", print_fields_str, PRINT_AVEDISKREAD, 0},
	{14, "AveDiskWrite", print_fields_str, PRINT_AVEDISKWRITE, 0},
	{10, "AvePages", print_fields_str, PRINT_AVEPAGES, 0},
	{10, "AveRSS", print_fields_str, PRINT_AVERSS, 0},
	{10, "AveVMSize", print_fields_str, PRINT_AVEVSIZE, 0},
	{16, "BlockID", print_fields_str, PRINT_BLOCKID, 0},
	{10, "Cluster", print_fields_str, PRINT_CLUSTER, 0},
	{14, "Comment", print_fields_str, PRINT_COMMENT, 0},
	{19, "Constraints", print_fields_str, PRINT_CONSTRAINTS, 0},
	{14, "ConsumedEnergy", print_fields_str, PRINT_CONSUMED_ENERGY, 0},
	{17, "ConsumedEnergyRaw", print_fields_uint64,
	 PRINT_CONSUMED_ENERGY_RAW, 0},
	{19, "Container", print_fields_str, PRINT_CONTAINER, 0},
	{10, "CPUTime", print_fields_time_from_secs, PRINT_CPU_TIME, 0},
	{10, "CPUTimeRAW", print_fields_uint64, PRINT_CPU_TIME_RAW, 0},
	{-10, "DBIndex", print_fields_uint64, PRINT_DB_INX, 0},
	{15, "DerivedExitCode", print_fields_str, PRINT_DERIVED_EC, 0},
	{10, "Elapsed", print_fields_time_from_secs, PRINT_ELAPSED, 0},
	{10, "ElapsedRaw", print_fields_uint32, PRINT_ELAPSED_RAW ,0},
	{19, "Eligible", print_fields_date, PRINT_ELIGIBLE ,0},
	{19, "End", print_fields_date, PRINT_END ,0},
	{8,  "ExitCode", print_fields_str, PRINT_EXITCODE, 0},
	{19, "Flags", print_fields_str, PRINT_FLAGS, 0},
	{6,  "GID", print_fields_uint, PRINT_GID, 0},
	{9,  "Group", print_fields_str, PRINT_GROUP, 0},
	{-12, "JobID", print_fields_str, PRINT_JOBID, 0},
	{-12, "JobIDRaw", print_fields_str, PRINT_JOBIDRAW, 0},
	{10, "JobName", print_fields_str, PRINT_JOBNAME, 0},
	{9,  "Layout", print_fields_str, PRINT_LAYOUT, 0},
	{12, "MaxDiskRead", print_fields_str, PRINT_MAXDISKREAD, 0},
	{15, "MaxDiskReadNode", print_fields_str, PRINT_MAXDISKREADNODE, 0},
	{15, "MaxDiskReadTask", print_fields_uint, PRINT_MAXDISKREADTASK, 0},
	{12, "MaxDiskWrite", print_fields_str, PRINT_MAXDISKWRITE, 0},
	{16, "MaxDiskWriteNode", print_fields_str, PRINT_MAXDISKWRITENODE, 0},
	{16, "MaxDiskWriteTask", print_fields_uint, PRINT_MAXDISKWRITETASK, 0},
	{8,  "MaxPages", print_fields_str, PRINT_MAXPAGES, 0},
	{12, "MaxPagesNode", print_fields_str, PRINT_MAXPAGESNODE, 0},
	{14, "MaxPagesTask", print_fields_uint, PRINT_MAXPAGESTASK, 0},
	{10, "MaxRSS", print_fields_str, PRINT_MAXRSS, 0},
	{10, "MaxRSSNode", print_fields_str, PRINT_MAXRSSNODE, 0},
	{10, "MaxRSSTask", print_fields_uint, PRINT_MAXRSSTASK, 0},
	{10, "MaxVMSize", print_fields_str, PRINT_MAXVSIZE, 0},
	{14, "MaxVMSizeNode", print_fields_str, PRINT_MAXVSIZENODE, 0},
	{14, "MaxVMSizeTask", print_fields_uint, PRINT_MAXVSIZETASK, 0},
	{12, "McsLabel", print_fields_str, PRINT_MCS_LABEL, 0},
	{10, "MinCPU", print_fields_str, PRINT_MINCPU, 0},
	{10, "MinCPUNode", print_fields_str, PRINT_MINCPUNODE, 0},
	{10, "MinCPUTask", print_fields_uint, PRINT_MINCPUTASK, 0},
	{10, "NCPUS", print_fields_uint, PRINT_ALLOC_CPUS, 0},
	{8,  "NNodes", print_fields_uint, PRINT_NNODES, 0},
	{15, "NodeList", print_fields_str, PRINT_NODELIST, 0},
	{8,  "NTasks", print_fields_uint, PRINT_NTASKS, 0},
	{10, "Partition", print_fields_str, PRINT_PARTITION, 0},
	{10, "Priority", print_fields_uint, PRINT_PRIO, 0},
	{10, "QOS", print_fields_str, PRINT_QOS, 0},
	{6,  "QOSRAW", print_fields_uint, PRINT_QOSRAW, 0},
	{22, "Reason", print_fields_str, PRINT_REASON, 0},
	{10, "ReqCPUFreq", print_fields_str, PRINT_REQ_CPUFREQ_MAX, 0}, /* vestigial */
	{13, "ReqCPUFreqGov", print_fields_str, PRINT_REQ_CPUFREQ_GOV, 0},
	{13, "ReqCPUFreqMax", print_fields_str, PRINT_REQ_CPUFREQ_MAX, 0},
	{13, "ReqCPUFreqMin", print_fields_str, PRINT_REQ_CPUFREQ_MIN, 0},
	{8,  "ReqCPUS", print_fields_uint, PRINT_REQ_CPUS, 0},
	{10, "ReqMem", print_fields_str, PRINT_REQ_MEM, 0},
	{8,  "ReqNodes", print_fields_str, PRINT_REQ_NODES, 0},
	{10, "ReqTRES", print_fields_str, PRINT_TRESR, 0},
	{20, "Reservation",  print_fields_str, PRINT_RESERVATION, 0},
	{8,  "ReservationId",  print_fields_uint, PRINT_RESERVATION_ID, 0},
	{10, "Reserved", print_fields_time_from_secs, PRINT_RESV, 0},
	{10, "ResvCPU", print_fields_time_from_secs, PRINT_RESV_CPU, 0},
	{10, "ResvCPURAW", print_fields_uint, PRINT_RESV_CPU, 0},
	{19, "Start", print_fields_date, PRINT_START, 0},
	{10, "State", print_fields_str, PRINT_STATE, 0},
	{19, "Submit", print_fields_date, PRINT_SUBMIT, 0},
	{20, "SubmitLine", print_fields_str, PRINT_SUBMIT_LINE,0},
	{10, "Suspended", print_fields_time_from_secs, PRINT_SUSPENDED, 0},
	{15, "SystemComment", print_fields_str, PRINT_SYSTEM_COMMENT, 0},
	{10, "SystemCPU", print_fields_str, PRINT_SYSTEMCPU, 0},
	{10, "Timelimit", print_fields_str, PRINT_TIMELIMIT, 0},
	{10, "TimelimitRaw", print_fields_str, PRINT_TIMELIMIT_RAW, 0},
	{10, "TotalCPU", print_fields_str, PRINT_TOTALCPU, 0},
	{14, "TRESUsageInAve", print_fields_str, PRINT_TRESUIA, 0},
	{14, "TRESUsageInMax", print_fields_str, PRINT_TRESUIM},
	{18, "TRESUsageInMaxNode", print_fields_str, PRINT_TRESUIMN, 0},
	{18, "TRESUsageInMaxTask", print_fields_str, PRINT_TRESUIMT, 0},
	{14, "TRESUsageInMin", print_fields_str, PRINT_TRESUIMI, 0},
	{18, "TRESUsageInMinNode", print_fields_str, PRINT_TRESUIMIN, 0},
	{18, "TRESUsageInMinTask", print_fields_str, PRINT_TRESUIMIT, 0},
	{14, "TRESUsageInTot", print_fields_str, PRINT_TRESUIT, 0},
	{15, "TRESUsageOutAve", print_fields_str, PRINT_TRESUOA, 0},
	{15, "TRESUsageOutMax", print_fields_str, PRINT_TRESUOM, 0},
	{19, "TRESUsageOutMaxNode", print_fields_str, PRINT_TRESUOMN, 0},
	{19, "TRESUsageOutMaxTask", print_fields_str, PRINT_TRESUOMT, 0},
	{15, "TRESUsageOutMin", print_fields_str, PRINT_TRESUOMI, 0},
	{19, "TRESUsageOutMinNode", print_fields_str, PRINT_TRESUOMIN, 0},
	{19, "TRESUsageOutMinTask", print_fields_str, PRINT_TRESUOMIT, 0},
	{15, "TRESUsageOutTot", print_fields_str, PRINT_TRESUOT, 0},
	{6,  "UID", print_fields_uint, PRINT_UID, 0},
	{9,  "User", print_fields_str, PRINT_USER, 0},
	{10, "UserCPU", print_fields_str, PRINT_USERCPU, 0},
	{10, "WCKey", print_fields_str, PRINT_WCKEY, 0},
	{10, "WCKeyID", print_fields_uint, PRINT_WCKEYID, 0},
	{20, "WorkDir", print_fields_str, PRINT_WORK_DIR, 0},
#ifdef __METASTACK_OPT_RESC_NODEDETAIL
	{40, "ResourceNodeDetail", print_fields_str, PRINT_RESCNODEDETAIL, 0},
#endif
#ifdef __METASTACK_OPT_SACCT_COMMAND
	{32, "Command", print_fields_str, PRINT_COMMAND, 0},
#endif
#ifdef __METASTACK_OPT_SACCT_OUTPUT
	{32, "Stdout", print_fields_str, PRINT_STDOUT, 0},
	{32, "Stderr", print_fields_str, PRINT_STDERR, 0},
#endif
	{0,  NULL, NULL, 0, 0}
};
#else
print_field_t fields[] = {
	{10, "Account", print_fields_str, PRINT_ACCOUNT},
	{15, "AdminComment", print_fields_str, PRINT_ADMIN_COMMENT},
	{10, "AllocCPUS", print_fields_uint, PRINT_ALLOC_CPUS},
	{10, "AllocNodes", print_fields_str, PRINT_ALLOC_NODES},
	{10, "AllocTRES", print_fields_str, PRINT_TRESA},
	{7,  "AssocID", print_fields_uint, PRINT_ASSOCID},
	{10, "AveCPU", print_fields_str, PRINT_AVECPU},
	{10, "AveCPUFreq", print_fields_str, PRINT_ACT_CPUFREQ},
	{14, "AveDiskRead", print_fields_str, PRINT_AVEDISKREAD},
	{14, "AveDiskWrite", print_fields_str, PRINT_AVEDISKWRITE},
	{10, "AvePages", print_fields_str, PRINT_AVEPAGES},
	{10, "AveRSS", print_fields_str, PRINT_AVERSS},
	{10, "AveVMSize", print_fields_str, PRINT_AVEVSIZE},
	{16, "BlockID", print_fields_str, PRINT_BLOCKID},
	{10, "Cluster", print_fields_str, PRINT_CLUSTER},
	{14, "Comment", print_fields_str, PRINT_COMMENT},
	{19, "Constraints", print_fields_str, PRINT_CONSTRAINTS},
	{14, "ConsumedEnergy", print_fields_str, PRINT_CONSUMED_ENERGY},
	{17, "ConsumedEnergyRaw", print_fields_uint64,
	 PRINT_CONSUMED_ENERGY_RAW},
	{19, "Container", print_fields_str, PRINT_CONTAINER},
	{10, "CPUTime", print_fields_time_from_secs, PRINT_CPU_TIME},
	{10, "CPUTimeRAW", print_fields_uint64, PRINT_CPU_TIME_RAW},
	{-10, "DBIndex", print_fields_uint64, PRINT_DB_INX},
	{15, "DerivedExitCode", print_fields_str, PRINT_DERIVED_EC},
	{10, "Elapsed", print_fields_time_from_secs, PRINT_ELAPSED},
	{10, "ElapsedRaw", print_fields_uint32, PRINT_ELAPSED_RAW},
	{19, "Eligible", print_fields_date, PRINT_ELIGIBLE},
	{19, "End", print_fields_date, PRINT_END},
	{8,  "ExitCode", print_fields_str, PRINT_EXITCODE},
	{19, "Flags", print_fields_str, PRINT_FLAGS},
	{6,  "GID", print_fields_uint, PRINT_GID},
	{9,  "Group", print_fields_str, PRINT_GROUP},
	{-12, "JobID", print_fields_str, PRINT_JOBID},
	{-12, "JobIDRaw", print_fields_str, PRINT_JOBIDRAW},
	{10, "JobName", print_fields_str, PRINT_JOBNAME},
	{9,  "Layout", print_fields_str, PRINT_LAYOUT},
	{12, "MaxDiskRead", print_fields_str, PRINT_MAXDISKREAD},
	{15, "MaxDiskReadNode", print_fields_str, PRINT_MAXDISKREADNODE},
	{15, "MaxDiskReadTask", print_fields_uint, PRINT_MAXDISKREADTASK},
	{12, "MaxDiskWrite", print_fields_str, PRINT_MAXDISKWRITE},
	{16, "MaxDiskWriteNode", print_fields_str, PRINT_MAXDISKWRITENODE},
	{16, "MaxDiskWriteTask", print_fields_uint, PRINT_MAXDISKWRITETASK},
	{8,  "MaxPages", print_fields_str, PRINT_MAXPAGES},
	{12, "MaxPagesNode", print_fields_str, PRINT_MAXPAGESNODE},
	{14, "MaxPagesTask", print_fields_uint, PRINT_MAXPAGESTASK},
	{10, "MaxRSS", print_fields_str, PRINT_MAXRSS},
	{10, "MaxRSSNode", print_fields_str, PRINT_MAXRSSNODE},
	{10, "MaxRSSTask", print_fields_uint, PRINT_MAXRSSTASK},
	{10, "MaxVMSize", print_fields_str, PRINT_MAXVSIZE},
	{14, "MaxVMSizeNode", print_fields_str, PRINT_MAXVSIZENODE},
	{14, "MaxVMSizeTask", print_fields_uint, PRINT_MAXVSIZETASK},
	{12, "McsLabel", print_fields_str, PRINT_MCS_LABEL},
	{10, "MinCPU", print_fields_str, PRINT_MINCPU},
	{10, "MinCPUNode", print_fields_str, PRINT_MINCPUNODE},
	{10, "MinCPUTask", print_fields_uint, PRINT_MINCPUTASK},
	{10, "NCPUS", print_fields_uint, PRINT_ALLOC_CPUS},
	{8,  "NNodes", print_fields_uint, PRINT_NNODES},
	{15, "NodeList", print_fields_str, PRINT_NODELIST},
	{8,  "NTasks", print_fields_uint, PRINT_NTASKS},
	{10, "Partition", print_fields_str, PRINT_PARTITION},
	{10, "Priority", print_fields_uint, PRINT_PRIO},
	{10, "QOS", print_fields_str, PRINT_QOS},
	{6,  "QOSRAW", print_fields_uint, PRINT_QOSRAW},
	{22, "Reason", print_fields_str, PRINT_REASON},
	{10, "ReqCPUFreq", print_fields_str, PRINT_REQ_CPUFREQ_MAX}, /* vestigial */
	{13, "ReqCPUFreqGov", print_fields_str, PRINT_REQ_CPUFREQ_GOV},
	{13, "ReqCPUFreqMax", print_fields_str, PRINT_REQ_CPUFREQ_MAX},
	{13, "ReqCPUFreqMin", print_fields_str, PRINT_REQ_CPUFREQ_MIN},
	{8,  "ReqCPUS", print_fields_uint, PRINT_REQ_CPUS},
	{10, "ReqMem", print_fields_str, PRINT_REQ_MEM},
	{8,  "ReqNodes", print_fields_str, PRINT_REQ_NODES},
	{10, "ReqTRES", print_fields_str, PRINT_TRESR},
	{20, "Reservation",  print_fields_str, PRINT_RESERVATION},
	{8,  "ReservationId",  print_fields_uint, PRINT_RESERVATION_ID},
	{10, "Reserved", print_fields_time_from_secs, PRINT_RESV},
	{10, "ResvCPU", print_fields_time_from_secs, PRINT_RESV_CPU},
	{10, "ResvCPURAW", print_fields_uint, PRINT_RESV_CPU},
	{19, "Start", print_fields_date, PRINT_START},
	{10, "State", print_fields_str, PRINT_STATE},
	{19, "Submit", print_fields_date, PRINT_SUBMIT},
	{20, "SubmitLine", print_fields_str, PRINT_SUBMIT_LINE},
	{10, "Suspended", print_fields_time_from_secs, PRINT_SUSPENDED},
	{15, "SystemComment", print_fields_str, PRINT_SYSTEM_COMMENT},
	{10, "SystemCPU", print_fields_str, PRINT_SYSTEMCPU},
	{10, "Timelimit", print_fields_str, PRINT_TIMELIMIT},
	{10, "TimelimitRaw", print_fields_str, PRINT_TIMELIMIT_RAW},
	{10, "TotalCPU", print_fields_str, PRINT_TOTALCPU},
	{14, "TRESUsageInAve", print_fields_str, PRINT_TRESUIA},
	{14, "TRESUsageInMax", print_fields_str, PRINT_TRESUIM},
	{18, "TRESUsageInMaxNode", print_fields_str, PRINT_TRESUIMN},
	{18, "TRESUsageInMaxTask", print_fields_str, PRINT_TRESUIMT},
	{14, "TRESUsageInMin", print_fields_str, PRINT_TRESUIMI},
	{18, "TRESUsageInMinNode", print_fields_str, PRINT_TRESUIMIN},
	{18, "TRESUsageInMinTask", print_fields_str, PRINT_TRESUIMIT},
	{14, "TRESUsageInTot", print_fields_str, PRINT_TRESUIT},
	{15, "TRESUsageOutAve", print_fields_str, PRINT_TRESUOA},
	{15, "TRESUsageOutMax", print_fields_str, PRINT_TRESUOM},
	{19, "TRESUsageOutMaxNode", print_fields_str, PRINT_TRESUOMN},
	{19, "TRESUsageOutMaxTask", print_fields_str, PRINT_TRESUOMT},
	{15, "TRESUsageOutMin", print_fields_str, PRINT_TRESUOMI},
	{19, "TRESUsageOutMinNode", print_fields_str, PRINT_TRESUOMIN},
	{19, "TRESUsageOutMinTask", print_fields_str, PRINT_TRESUOMIT},
	{15, "TRESUsageOutTot", print_fields_str, PRINT_TRESUOT},
	{6,  "UID", print_fields_uint, PRINT_UID},
	{9,  "User", print_fields_str, PRINT_USER},
	{10, "UserCPU", print_fields_str, PRINT_USERCPU},
	{10, "WCKey", print_fields_str, PRINT_WCKEY},
	{10, "WCKeyID", print_fields_uint, PRINT_WCKEYID},
	{20, "WorkDir", print_fields_str, PRINT_WORK_DIR},
#ifdef __METASTACK_OPT_RESC_NODEDETAIL
	{40, "ResourceNodeDetail", print_fields_str, PRINT_RESCNODEDETAIL},
#endif
#ifdef __METASTACK_OPT_SACCT_COMMAND
	{32, "Command", print_fields_str, PRINT_COMMAND},
#endif
#ifdef __METASTACK_OPT_SACCT_OUTPUT
	{32, "Stdout", print_fields_str, PRINT_STDOUT},
	{32, "Stderr", print_fields_str, PRINT_STDERR},
#endif
	{0,  NULL, NULL, 0}
};
#endif
List jobs = NULL;

int main(int argc, char **argv)
{
	enum {
		SACCT_LIST,
		SACCT_LIST_DATA,
		SACCT_HELP,
		SACCT_USAGE
	} op;
	int rc = 0;

	slurm_conf_init(NULL);
	sacct_init();
	parse_command_line(argc, argv);

	/* What are we doing? Requests for help take highest priority,
	 * but then check for illogical switch combinations.
	 */

	if (params.opt_help)
		op = SACCT_HELP;
	else if (params.mimetype)
		op = SACCT_LIST_DATA;
	else
		op = SACCT_LIST;


	switch (op) {
	case SACCT_LIST:
		if (!(params.job_cond->flags & JOBCOND_FLAG_SCRIPT) &&
		    !(params.job_cond->flags & JOBCOND_FLAG_ENV))
			print_fields_header(print_fields_list);
		if (get_data() == SLURM_ERROR)
			exit(errno);
		if (params.opt_completion)
			do_list_completion();
		else
			do_list();
		break;
	case SACCT_LIST_DATA:
		dump_data(argc, argv);
		break;
	case SACCT_HELP:
		do_help();
		break;
	default:
		fprintf(stderr, "sacct bug: should never get here\n");
		sacct_fini();
		exit(2);
	}

	sacct_fini();

#if MEMORY_LEAK_DEBUG
	data_fini();
	slurm_auth_fini();
	slurm_conf_destroy();
	log_fini();
#endif

	return (rc);
}
